controllers/etcdcluster_controller.go

```diff
@@ -86,51 +115,131 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=core,resources=endpoints,verbs=get;list
+
+func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	cluster := &etcdv1alpha1.EtcdCluster{}
+	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+		log.Error(err, "unable to fetch EtcdCluster resource")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+		// special 'exit early' case.
+	}
+
+	// Attempt to dial the etcd cluster, recording the cluster response if we can
+	members, err := r.connectToEtcd(ctx, cluster)
+	if err != nil {
+		log.Error(err, "Unable to contact etcd cluster")
+	}
+
+	// Perform a reconcile, getting back the desired result, any errors, and a clusterEvent. This is an internal concept
+	// and is not the same as the Kubernetes event, although it is used to make one later.
+	result, clusterEvent, err := r.reconcile(ctx, req.NamespacedName, members, cluster)
+	if err != nil {
+		log.Error(err, "Failed to reconcile")
+	}
+
+	// The update status takes in the cluster definition, and the member list from etcd as of *before we ran reconcile*.
+	// We also get the event, which may contain rich information about what we did (such as the new member name on a
+	// MemberAdded event).
+	err = r.updateStatus(ctx, req.NamespacedName, cluster, members, clusterEvent)
+	if err != nil {
+		log.Error(err, "Failed to update status")
+	}
+
+	// Finally, the event is used to generate a Kubernetes event by calling `Record` and passing in the recorder.
+	if clusterEvent != nil {
+		clusterEvent.Record(r.Recorder)
+	}
+
+	return result, nil
+}
+
+func (r *EtcdClusterReconciler) connectToEtcd(ctx context.Context, cluster *etcdv1alpha1.EtcdCluster) ([]etcdclient.Member, error) {
+	etcdConfig := etcdClientConfig(cluster)
+	c, err := etcdclient.New(*etcdConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to connect to etcd")
+	}
+
+	membersAPI := etcdclient.NewMembersAPI(c)
+	members, err := membersAPI.List(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to list members of etcd cluster")
+	}
+
+	return members, nil
+}
+
+func etcdClientConfig(cluster *etcdv1alpha1.EtcdCluster) *etcdclient.Config {
+	peerNames := expectedPeerNamesForCluster(cluster)
+	peerAddresses := make([]string, len(peerNames))
+	for i, peerName := range peerNames {
+		u := &url.URL{
+			Scheme: etcdScheme,
+			Host:   fmt.Sprintf("%s:%d", expectedURLForPeer(cluster, peerName), etcdClientPort),
+		}
+		peerAddresses[i] = u.String()
```

Comment:
what's the reason for dialling the pods directly, rather than through the headless service?

Reply:
We construct the expected DNS names which uses the headless service. The other options would be to pull the `Endpoints` and use those. This doesn't actually work because you end up trying to talk to the peers using a IP and not a host and the peer doesn't seem to like it. (As they want you to use their advertise URLs).

This way constructs the expected advertise URLs, and just connects to those.

Reply:
but can we access through a single endpoint `my-cluster.default.svc:2379` ? like we do in the example in the README

Reply:
Will that resolve correctly? If so then yeah lets just do that. If you DNS for `servicename.namespace.svc` you'll get n `A` records, one for each peer. I know `etcdctl` can handle this, can the library client? (I'm testing it).

Reply:
It does seem to work.